src/dos/drive_local.cpp

```diff
@@ -697,65 +622,54 @@ bool localFile::Write(uint8_t *data, uint16_t *size)
 	}
 
 	// Otherwise we have some data to write
-	const auto requested = *size;
-	const auto actual = static_cast<uint16_t>(fwrite(data, 1, requested, fhandle));
-	if (actual != requested) {
-		LOG_DEBUG("FS: Only wrote %u of %u requested bytes to file '%s'",
-		          actual,
-		          requested,
-		          name.c_str());
-
-		// Check for host write error
-		if (ferror(fhandle)) {
-			clearerr(fhandle);
-			DOS_SetError(DOSERR_ACCESS_DENIED);
-			return false;
-		}
+	const auto ret = write_native_file(file_handle, data, *size);
+	*size          = check_cast<uint16_t>(ret.num_bytes);
+	if (ret.error) {
+		DOS_SetError(DOSERR_ACCESS_DENIED);
+		return false;
 	}
-	*size = actual; // always save the actual
-	return true;    // always return true, even if partially written
+
+	return true;
 }
 
 bool localFile::Seek(uint32_t *pos_addr, uint32_t type)
 {
-	int seektype;
+	NativeSeek seek_type;
 	switch (type) {
-	case DOS_SEEK_SET:seektype=SEEK_SET;break;
-	case DOS_SEEK_CUR:seektype=SEEK_CUR;break;
-	case DOS_SEEK_END:seektype=SEEK_END;break;
-	default:
-	//TODO Give some doserrorcode;
-		return false;//ERROR
+	case DOS_SEEK_SET: seek_type = NativeSeek::Set; break;
+	case DOS_SEEK_CUR: seek_type = NativeSeek::Current; break;
+	case DOS_SEEK_END: seek_type = NativeSeek::End; break;
+	default: assertm(false, "Invalid seek type"); return false;
 	}
 
 	// The inbound position is actually an int32_t being passed through a
 	// uint32_t* pointer (pos_addr), so reinterpret the underlying memory as
 	// such to prevent rollover into the unsigned range.
-	const auto pos = *reinterpret_cast<int32_t *>(pos_addr);
-	if (!fseek_to_and_check(pos, seektype)) {
+	const auto requested_pos = *reinterpret_cast<int32_t*>(pos_addr);
+
+	auto returned_pos = seek_native_file(file_handle, requested_pos, seek_type);
+
+	// TODO: Test Black Throne and see if this hack is still needed
```

Comment:
Could you test this before merging?

Reply:
Tested.  It worked with this hack in and failed to launch with it removed.  I pushed a new commit with more accurate/less hacky behavior and tested this works on Linux and Windows.

I was side-eyeing that `reinterpret_cast` and it turned out to be incorrect for `SEEK_SET`.  The game seeks to 2777296400 (over 2GB) into the file.  Windows and POSIX both allow this.  It doesn't actually resize the file on seek but if you then write to the file, it will fill with zeros up to that size.

Anyway, we just need to interpret the value as unsigned if using `SEEK_SET`.  I put some references to MS-DOS Encyclopedia in the comments.

The error handling and set to 0 isn't needed by Blackthorne but may be needed by other DOS games doing weird things.  Apparently the DOS Seek interrupt never fails assuming you give it a valid handle and seek type.

Reply:
Thanks for the detailed explanation @weirddan455. This seems to have been quite a rabbit hole, but I'm glad you got to the bottom of it 🎉

Reply:
It didn't actually take me too long.  Luckily it was an easily reproducible error and fail to launch on Blackthorne.  Then I just added a log statement and saw it was being cast to a large negative seek value, where it should have always been positive.